src/frontend/src/flows/authenticate/index.ts

```diff
@@ -0,0 +1,363 @@
+            return handleAuthSuccess(result, authContext);
+          }
+          // something went wrong, display error and try again
+          return displayError({
+            title: result.title,
+            message: result.message,
+            detail: result.detail !== "" ? result.detail : undefined,
+            primaryButton: "Try again",
+          }).then(() => init(authContext, editMode, userNumber));
+        })
+        .then(resolve);
+    };
+  });
+};
+
+const authenticateUser = async (
+  authContext: AuthContext,
+  editMode: boolean
+): Promise<AuthSuccess | null> => {
+  const userNumber = readUserNumber(editMode);
+  try {
+    if (userNumber === undefined) {
+      toggleErrorMessage("userNumberInput", "invalidAnchorMessage", true);
+      return null;
+    }
+    const result = await withLoader(() => IIConnection.login(userNumber));
+    const loginResult = apiResultToLoginFlowResult(result);
+    if (loginResult.tag === "ok") {
+      return await handleAuthSuccess(loginResult, authContext);
+    }
+    await displayError({
+      title: loginResult.title,
+      message: loginResult.message,
+      detail: loginResult.detail !== "" ? loginResult.detail : undefined,
+      primaryButton: "Try again",
+    });
+  } catch (error) {
+    await displayError({
+      title: "Authentication Failed",
+      message: "An error occurred during authentication.",
+      detail: error instanceof Error ? error.message : JSON.stringify(error),
+      primaryButton: "Try again",
+    });
+  }
+  return init(authContext, editMode, userNumber);
+};
+
+const displayPage = (
+  origin: string,
+  editMode: boolean,
+  userNumber?: bigint
+) => {
+  const container = document.getElementById("pageContent") as HTMLElement;
+  render(pageContent(origin, editMode, userNumber), container);
+};
+
+async function handleAuthSuccess(
+  loginResult: LoginFlowSuccess,
+  authContext: AuthContext
+) {
+  // successful login, store user number for next time
+  setUserNumber(loginResult.userNumber);
+  const [userKey, parsed_signed_delegation] = await withLoader(() =>
+    fetchDelegation(loginResult, authContext)
+  );
+  return {
+    userNumber: loginResult.userNumber,
+    connection: loginResult.connection,
+    sendDelegationMessage: () =>
+      authContext.postMessageCallback({
+        kind: "authorize-client-success",
+        delegations: [parsed_signed_delegation],
+        userPublicKey: Uint8Array.from(userKey),
+      }),
+  };
+}
+
+const initRecovery = (editMode: boolean) => {
+  const recoverButton = document.getElementById(
+    "recoverButton"
+  ) as HTMLButtonElement;
+  recoverButton.onclick = () => {
+    const userNumber = readUserNumber(editMode);
+    if (userNumber !== undefined) {
+      return useRecovery(userNumber);
+    }
+    toggleErrorMessage("userNumberInput", "invalidAnchorMessage", true);
+  };
+};
+
+const redirectToWelcomeScreen = () => {
+  window.location.hash = "";
+  window.location.reload();
```

Comment:
nit: I would inline those 2 lines, it would make it clear that the promise is not just dropped and forgotten about, but that the page is thrashed proper